heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/YarnScheduler.java

```diff
@@ -84,12 +93,32 @@ public boolean onRestart(RestartTopologyRequest request) {
 
   @Override
   public boolean onUpdate(UpdateTopologyRequest request) {
-    LOG.severe("Topology onUpdate not implemented by this scheduler.");
-    return false;
+    try {
+      updateTopologyManager.updateTopology(
+          request.getCurrentPackingPlan(), request.getProposedPackingPlan());
+    } catch (ExecutionException | InterruptedException e) {
+      LOG.log(Level.SEVERE, "Could not update topology for request: " + request, e);
+      return false;
+    }
+    return true;
   }
 
   @Override
   public void close() {
     HeronMasterDriverProvider.getInstance().killTopology();
   }
+
+  @Override
+  public void addContainers(Set<PackingPlan.ContainerPlan> containersToAdd) {
+    try {
+      HeronMasterDriverProvider.getInstance().scheduleHeronWorkers(containersToAdd);
+    } catch (HeronMasterDriver.ContainerAllocationException e) {
```

Comment:
As a follow-up, we should move `ContainerAllocationException` into spi and make make `IScalable` throw it. Same thing for `ContainerRemovalException`.